scio-jdbc/src/main/java/com/spotify/scio/jdbc/sharded/JdbcShardedSource.java

```diff
@@ -0,0 +1,230 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.jdbc.sharded;
+
+import com.spotify.scio.jdbc.JdbcConnectionOptions;
+import com.spotify.scio.jdbc.sharded.JdbcShardedReadOptions;
+import com.spotify.scio.jdbc.sharded.Range;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class JdbcShardedSource<T> extends BoundedSource<T> {
+
+  private static final Logger log = LoggerFactory.getLogger(JdbcShardedSource.class);
+
+  private final Range<Long> range;
+  private final Coder<T> coder;
+  private final JdbcShardedReadOptions<T> readOptions;
+
+  public JdbcShardedSource(
+      JdbcShardedReadOptions<T> readOptions, Coder<T> coder, Range<Long> range) {
+    this.readOptions = readOptions;
+    this.range = range;
+    this.coder = coder;
+  }
+
+  private static Connection createConnection(JdbcConnectionOptions connectionOptions)
+      throws Exception {
+    final Connection connection =
+        DriverManager.getConnection(
+            connectionOptions.connectionUrl(),
+            connectionOptions.username(),
+            connectionOptions.password().get());
+    connection.setAutoCommit(false);
+    log.info("Created connection to [{}]", connectionOptions.connectionUrl());
+
+    return connection;
+  }
+
+  private List<Range<Long>> partitionRange(Range<Long> range, int partitionsCount) {
+    final long partitionSize = (range.max() - range.min()) / partitionsCount;
+
+    return Stream.iterate(0, i -> i + 1)
+        .limit(partitionsCount)
+        .map(
+            i ->
+                (i < partitionsCount - 1)
+                    ? new Range<Long>(
+                        range.min() + i * partitionSize,
+                        range.min() + i * partitionSize + partitionSize - 1)
+                    : new Range<Long>(range.min() + i * partitionSize, range.max()))
+        .collect(Collectors.toList());
+  }
+
+  private Range<Long> getShardColumnRange() throws Exception {
+    try (Connection connection = createConnection(readOptions.connectionOptions())) {
+      String query =
```

Comment:
A bit on the defensive side: should we also check if the column has index?

Reply:
I think it should be doable with [DatabaseMetaData.getIndexInfo](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getIndexInfo-java.lang.String-java.lang.String-java.lang.String-boolean-boolean-). There is one caveat though - a column could be in an index, but the index might be inefficient e.g. a composite index where the first column has low cardinality. But we can add this check just to filter out cases when the column definitely not in an index.